src/Network/GRPC/MQTT/Message/Packet.hs

```diff
@@ -70,129 +62,67 @@ import Relude
 import UnliftIO (MonadUnliftIO)
 import UnliftIO.Async (replicateConcurrently_)
 
----------------------------------------------------------------------------------
-
-import Control.Concurrent.TMap (TMap)
-import Control.Concurrent.TMap qualified as TMap
-
-import Network.GRPC.MQTT.Message.Packet.Core
-  ( Packet (Packet, metadata, payload),
-    PacketInfo (PacketInfo, npackets, position),
-  )
-import Network.GRPC.MQTT.Message.TH (reifyFieldNumber, reifyRecordField)
-
-import Network.GRPC.MQTT.Message (WireDecodeError)
-import Network.GRPC.MQTT.Message qualified as Message
-
-import Network.GRPC.MQTT.Serial (WireDecodeOptions, WireEncodeOptions)
-
-import Proto3.Wire.Decode.Extra qualified as Decode
-import Proto3.Wire.Types.Extra (RecordField)
-
--- Packet - Splitting -----------------------------------------------------------
+-- Packet ----------------------------------------------------------------------
 
--- | @'splitPackets' n bytes@ splits the 'ByteString' @bytes@ into a vector of
--- packets with a 'ByteString' payload of length @n@-bytes.
---
--- A singleton packet vector with payload @bytes@ is produced in the case that
--- @n@ is negative or @bytes@ is an empty 'ByteString'.
+-- | TODO
 --
 -- @since 0.1.0.0
-splitPackets :: Int -> ByteString -> Vector (Packet ByteString)
-splitPackets size bytes
-  | size <= 0 || ByteString.null bytes =
-    -- In the case that the maximum packet payload length @size@ is not at least
-    -- 1 byte or the provided 'ByteString' @bytes@ is empty, then yield one
-    -- terminal packet.
-    Vector.singleton (Packet bytes (PacketInfo 0 1))
-  | otherwise = Vector.create do
-    -- Allocates a vector of containing @ByteString.length bytes / size@
-    -- elements rounded upwards. For example, fix the length of @bytes@ to
-    -- be 5 and @size@ to be 2, then a vector of length 3 is allocated.
-    let count :: Int
-        count = quotCeil (ByteString.length bytes) size
-     in MVector.generate count (makePacket count)
-  where
-    -- Takes the i-th @size@ length chunk of the 'ByteString' @bytes@.
-    sliceBytes :: Int -> ByteString
-    sliceBytes i = ByteString.take size (ByteString.drop (i * size) bytes)
-
-    -- 'quot' on 'Int', rounding upwards/toward positive infinity.
-    quotCeil :: Int -> Int -> Int
-    quotCeil a b = (a + (b - 1)) `quot` b
-
-    makePacket :: Int -> Int -> Packet ByteString
-    makePacket count i = Packet (sliceBytes i) (PacketInfo i count)
+data Packet msg = Packet
+  { payload :: msg
+  , position :: {-# UNPACK #-} !Int
+  , npackets :: {-# UNPACK #-} !Int
```

Comment:
Here I've inlined `PacketInfo`. My motivations for doing so: `PacketInfo` was never used in a situation where it wasn't a field of a `Packet` and to simplify/optimize serialization of `Packet`.